Improve color handling and enhance ColorPicker

CamilleGuillory opened the pull request:
### Why This Change?

The previous appearance customization logic (favicon, org banner, headers) was scattered and tightly coupled within [button.js](cci:7://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/button.js:0:0-0:0), making it difficult to maintain and extend. Org detection was brittle, leading to issues where **Production** banners might appear incorrectly or not at all if local storage was incomplete. There was also no centralized way to manage color settings or share utilities between the content script and extension options.

This PR refactors the appearance logic into a modular architecture, introducing dedicated utilities for color management, settings storage, and cross-context messaging. It also improves the robustness of org detection and banner rendering.

---

### Architecture Overview

The code has been reorganized into specialized modules:

#### **Domain Logic ([addon/utils/colorUtils.js](cci:7://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/utils/colorUtils.js:0:0-0:0))**
- Centralizes color palettes with environment classification (`dev`, `uat`, `int`, `full`) via `COLORS` and `COLOR_SHADES`.
- Provides robust color normalization through [colorNameToHex](cci:1://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/utils/colorUtils.js:208:0-218:1) and [normalizeColorEntry](cci:1://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/utils/colorUtils.js:189:0-194:1) (handling named colors, hex codes).
- Implements [getEnvironmentType(sfHost)](cci:1://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/utils/colorUtils.js:220:0-229:1) to infer environment from hostname patterns.
- Exports HSV/Hex conversion utilities ([hexToHsv](cci:1://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/utils/colorUtils.js:231:0-259:1), [hsvToHex](cci:1://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/utils/colorUtils.js:261:0-269:1)).
- Defines `DEFAULT_COLOR` (`#FF0000`) as the fallback.

#### **State Management ([addon/utils/colorSettingsStore.js](cci:7://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/utils/colorSettingsStore.js:0:0-0:0))**
- Manages reading/writing of global and org-specific color settings via [getColorSettingsSnapshot](cci:1://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/utils/colorSettingsStore.js:99:0-104:1) and [updateColorSettings](cci:1://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/utils/colorSettingsStore.js:106:0-116:1).
- Implements a subscription pattern ([subscribeToColorSettings](cci:1://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/utils/colorSettingsStore.js:118:0-126:1)) for reactive updates with [notifyColorSubscribers](cci:1://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/utils/colorSettingsStore.js:58:0-64:1).
- Provides smart color suggestions via [getSuggestedFaviconColor](cci:1://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/utils/colorSettingsStore.js:136:0-139:1) and [ensureFaviconColor](cci:1://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/utils/colorSettingsStore.js:141:0-148:1) based on environment type.
- Exports [getKnownOrgHosts()](cci:1://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/utils/colorSettingsStore.js:128:0-134:1) for discovering configured orgs.

#### **Communication Layer ([addon/utils/messaging.js](cci:7://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/utils/messaging.js:0:0-0:0))**
- Standardizes `chrome.runtime.sendMessage` and `postMessage` calls via [broadcastSettings](cci:1://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/utils/messaging.js:23:0-49:1).
- Includes security validation via [isTrustedSettingsMessage](cci:1://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/utils/messaging.js:51:0-55:1) with origin checking.
- Provides [deriveHostOrigin](cci:1://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/utils/messaging.js:15:0-21:1) and [getExtensionOrigin](cci:1://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/utils/messaging.js:13:0-13:57) utilities.
- Exports [addSettingsMessageListener](cci:1://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/utils/messaging.js:57:0-62:1) for easy event subscription.

#### **Consumer Implementation ([addon/button.js](cci:7://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/button.js:0:0-0:0))**
- Refactored to use the new utilities.
- Introduced [buildAppearanceContext(sfHost)](cci:1://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/button.js:188:2-212:3) to deterministically calculate the state (sandbox, trial, production, unknown) before applying visual changes.
- Added [getAppearanceFlags()](cci:1://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/button.js:214:2-219:3) for centralized reading of boolean toggle states (`colorizeFavicon`, `colorizeOrgBanner`).
- Consolidated DOM selectors into a `SELECTORS` constant for better maintainability.
- Simplified [observeElement](cci:1://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/button.js:166:2-183:3) implementation with early return pattern.

#### **Additional Consumer Updates**
- **[addon/options.js](cci:7://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/options.js:0:0-0:0)**: Uses [updateColorSettings](cci:1://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/utils/colorSettingsStore.js:106:0-116:1), [subscribeToColorSettings](cci:1://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/utils/colorSettingsStore.js:118:0-126:1), [getColorSettingsSnapshot](cci:1://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/utils/colorSettingsStore.js:99:0-104:1), and [broadcastSettings](cci:1://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/utils/messaging.js:23:0-49:1) for settings management.
- **[addon/popup.js](cci:7://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/popup.js:0:0-0:0)**: Listens for `settingsUpdated` runtime messages and forwards to parent window via [broadcastSettings](cci:1://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/utils/messaging.js:23:0-49:1).
- **`addon/components/ColorPicker.js`**: Updated to use centralized color utilities.
- **`addon/components/PageHeader.js`**: Updated for consistency with new color system.
- **`addon/inspector.js`**: Minor updates for compatibility.

---

### Key Changes & Features

#### **1. Improved Org Detection & Banner Safety**
Org classification logic has been rewritten inside [buildAppearanceContext](cci:1://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/button.js:188:2-212:3):

```javascript
// Check if org info is available
const isUnknown = !(isSandboxKey in iFrameLocalStorage);

let isSandbox = false;
let isProdOnly = false;

if (!isUnknown) {
  isSandbox = toBoolean(iFrameLocalStorage[isSandboxKey]);
  isProdOnly = !isSandbox && !hasTrial;
}
// If unknown, both isSandbox and isProdOnly remain false → no banner shown
```

- **Fix:** Prevents showing *"WARNING: THIS IS PRODUCTION"* when org info is **unknown** rather than confirmed Production.
- **Logic:** Explicit checks on `isSandbox` and `trialExpirationDate` flags with an `isUnknown` guard.

#### **2. Reusable Appearance Logic**
- [generateColorSvg(color)](cci:1://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/button.js:221:2-224:3): Extracted SVG generator for dynamic favicons.
- [getAppearanceFlags()](cci:1://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/button.js:214:2-219:3): Centralized reading of boolean toggle states.
- [applyAppearanceForHost(sfHost)](cci:1://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/button.js:238:2-254:3): Coordinates all visual updates.

#### **3. Favicon Revert Logic**
Added [revertFavicon(isExtensionPage)](cci:1://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/button.js:226:2-233:3) to restore the original page favicon when the feature is disabled or settings are cleared. Tracks `originalFaviconHref` for clean rollback:

```javascript
let originalFaviconHref = null;

function revertFavicon(isExtensionPage) {
  const current = document.querySelector(SELECTORS.FAVICON);
  if (!current || !current.dataset.sfInspectorFavicon) return;
  current.href = originalFaviconHref || (isExtensionPage ? "icon128.png" : "/favicon.ico");
  delete current.dataset.sfInspectorFavicon;
}
```

#### **4. Code Cleanup**
- [observeElement](cci:1://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/button.js:166:2-183:3): Simplified via early-return structure.
- `SELECTORS`: All DOM queries consolidated to eliminate repeated string literals.
- [handleSettingsUpdate](cci:1://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/button.js:407:4-431:5): Centralized message handling for settings updates from popup/options.

---

### Data Flow

1. **Context Building:**  
   [buildAppearanceContext(sfHost)](cci:1://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/button.js:188:2-212:3) loads and assembles environment + local storage data.

2. **Flag Resolution:**  
   [getAppearanceFlags()](cci:1://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/button.js:214:2-219:3) determines active features.

3. **Application:**  
   [applyAppearanceForHost](cci:1://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/button.js:238:2-254:3) coordinates updates:
   - If Favicon enabled → [updateFaviconLink](cci:1://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/button.js:256:2-269:3) → generates/applies SVG via [generateColorSvg](cci:1://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/button.js:221:2-224:3).
   - If Banner enabled → [updateSandboxAppearance](cci:1://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/button.js:271:2-278:3) / [updateProdAppearance](cci:1://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/button.js:280:2-286:3) → DOM updates using `SELECTORS`.

---

### Data Flow: Updating Appearance Settings

When a user applies a new color or toggle in the Options page, the system propagates changes through three layers to ensure consistency across the Extension UI and the Salesforce Page.

```
User Action (Options Page)
│
▼ [Component] options.js handlers (handleToggleChange / persistColorUpdates)
│
├─► 1. updateColorSettings(sfHost, {key: value}) [addon/utils/colorSettingsStore.js]
│       │
│       ├──► Write to localStorage (Persistence)
│       │       └── Keys: "colorizeFavicon", "sfHost_customFavicon", "sfHost_orgBannerText", etc.
│       │
│       └──► notifyColorSubscribers()
│               └──► subscribeToColorSettings callbacks → setState() (Instant Options UI Re-render)
│
└─► 2. postLocalSettingsUpdate() [addon/options.js]
        └──► broadcastSettings() [addon/utils/messaging.js]
                │
                ▼ Chrome Runtime Message ("settingsUpdated")
                  Payload: { sfHost, settings: { ...snapshot } }
                        │
                        ▼ [Extension Context] popup.js
                              ├──► browser.runtime.onMessage listener
                              ├──► sessionStorage.removeItem("filteredStorage")
                              │       └── Forces next popup open to read fresh localStorage
                              │
                              └──► broadcastSettings(targetWindow=parent, targetOrigin="*")
                                        │
                                        ▼ Window PostMessage ("insextSettingsUpdate")
                                                │
                                                ▼ [Salesforce Page] button.js
                                                      ├──► handleSettingsUpdate()
                                                      │       ├──► Update iFrameLocalStorage cache
                                                      │       └──► applyAppearanceForHost()
                                                      │
                                                      ├──► buildAppearanceContext()
                                                      │       └── Resolves org state (Sandbox vs Prod vs Trial vs Unknown)
                                                      │
                                                      ├──► updateFaviconLink()
                                                      │       └──► generateColorSvg() → Replace <link rel="icon">
                                                      │
                                                      └──► updateSandboxAppearance() / updateProdAppearance()
                                                              └──► DOM Manipulation (Header colors, Warning Banners)
```

---

### Files Changed

| File | Change Type | Description |
|------|-------------|-------------|
| [addon/utils/colorUtils.js](cci:7://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/utils/colorUtils.js:0:0-0:0) | Added | Color palettes, normalization, environment detection, HSV/Hex conversion |
| [addon/utils/colorSettingsStore.js](cci:7://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/utils/colorSettingsStore.js:0:0-0:0) | Added | Settings state management with subscription pattern |
| [addon/utils/messaging.js](cci:7://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/utils/messaging.js:0:0-0:0) | Added | Cross-context messaging utilities |
| [addon/button.js](cci:7://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/button.js:0:0-0:0) | Modified | Refactored to use new utilities, added [buildAppearanceContext](cci:1://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/button.js:188:2-212:3), `SELECTORS`, [revertFavicon](cci:1://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/button.js:226:2-233:3) |
| [addon/options.js](cci:7://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/options.js:0:0-0:0) | Modified | Integrated with new settings store and messaging |
| [addon/popup.js](cci:7://file:///Users/cguillory/Mon%20Drive/99%20-%20Repos/1.%20Mine/Salesforce-Inspector-reloaded/addon/popup.js:0:0-0:0) | Modified | Added runtime message listener for settings forwarding |
| `addon/components/ColorPicker.js` | Modified | Updated to use centralized color utilities |
| `addon/components/PageHeader.js` | Modified | Updated for consistency |
| `addon/inspector.js` | Modified | Minor compatibility updates |

---

### Migration Notes

- **Storage Compatibility:** Uses existing localStorage keys. No data migration is required for users.
- **Input Compatibility:** Existing hex codes in user settings continue to work as before.
- **Visuals:** The default fallback color is `#FF0000`.

tprouvot reviewed (commented):
Thanks for the PR @CamilleGuillory  ! 
Here are few remarks I would like to discuss

CamilleGuillory commented on addon/utils/colorUtils.js:
```diff
@@ -0,0 +1,290 @@
+ * https://www.w3schools.com/tags/ref_colornames.asp
+ */
+
+const COLOR_NAME_TO_HEX = {
```
The picker does not automatically calculate the picker cursor positions from the hex value.
It stores the color as HSV so we need to calculate the related HSV from color name / hex input.

CamilleGuillory commented on addon/utils/colorUtils.js:
```diff
@@ -0,0 +1,290 @@
+
+// Color shades organized by environment type for favicon auto-population
+export const COLOR_SHADES = {
+  dev: [ //blue shades
```
Good idea

CamilleGuillory committed:
Refactor appearance customization logic with improved org detection and banner handling

- Extract appearance logic into reusable helper functions (buildAppearanceContext, getAppearanceFlags, generateColorSvg)
- Add proper org detection state handling to prevent showing banners when org info is unavailable
- Implement favicon revert functionality with original href tracking
- Consolidate selector definitions into SELECTORS constant
- Simplify observeElement implementation with early return pattern

tprouvot commented:
Hi @CamilleGuillory 
I have a very strange behaviour, could you reproduce it ?
The popup is closing right after its opening 

![2025-12-11_17-33-21 (1)](https://github.com/user-attachments/assets/ce3fe14f-e964-43b6-af0d-0ec552892aa6)

CamilleGuillory commented:
Hi @tprouvot 
No I cannot (and never saw that behavior).
Is it on all pages?
I also tested with 'official beta' running at the same time (but not with Main extension)

CamilleGuillory committed:
resolve conflict ->  Make XML class an exported class

CamilleGuillory committed:
Merge branch 'releaseCandidate' into feature/fix-picker-colorname-to-hexav2

tprouvot commented:
Is replaced by a new one
